File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    vector<vector<int>> changes = {{1,0}, {0,1}, {0,-1},{-1,0}};

    void visitIsland(vector<vector<char>>& grid, vector<vector<bool>>& visited, int i, int j) {
        if (visited[i][j]) return;
        
        visited[i][j] = true;
        
        for (vector<int> ch : changes) {
            int ni = i+ch[0], nj = j+ch[1];
            if (ni < grid.size() && nj < grid[0].size() && grid[ni][nj] == '1') {
                visitIsland(grid, visited, ni, nj);
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();

        vector<vector<bool>> visited(n, vector<bool>(m, false));

        int count = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    count++;
                    visitIsland(grid, visited, i, j);
                }
            }
        }

        return count;
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
from collections import deque

class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        '''
        brute force, for each land cell do bfs - dijkstra's algorithm.

        optimized : 

            do multi sourced bfs. start from treasure chests and keep doing level wise bfs until finished.

        '''
        
        INF = 2147483647

        q = deque([])

        n = len(grid)
        m = len(grid[0])

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 0:
                    q.append((i,j))

        diffs = ((0,1),(1,0),(0,-1),(-1,0))
        visited = [[False for _ in range(m)] for _ in range(n)]

        while len(q) != 0:
            x, y = q.popleft()
            visited[x][y] = 1

            for dx, dy in diffs:
                nx = x+dx
                ny = y+dy

                if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == INF and not visited[nx][ny]:
                    visited[nx][ny] = 1
                    grid[nx][ny] = grid[x][y]+1
                    q.append((nx,ny))


        